## 1. Review
***

### What is Random Forest Regression?

Random forest is a model that constructs many decision trees on varius sub-samples of the data and uses the average of the predictions made by each individual tree to make a prediction.

## 2. Random Forest Regression (without XGBoost)
***


In [1]:
import pandas as pd
import numpy as np
from sklearn import linear_model
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
train = pd.read_csv("../SalesPrediction/competitive-data-science-predict-future-sales/sales_train.csv")
train.head()

date  date_block_num  shop_id  item_id  item_price  item_cnt_day
0  02.01.2013               0       59    22154      999.00           1.0
1  03.01.2013               0       25     2552      899.00           1.0
2  05.01.2013               0       25     2552      899.00          -1.0
3  06.01.2013               0       25     2554     1709.05           1.0
4  15.01.2013               0       25     2555     1099.00           1.0

In [2]:
#Getting X and Y variables to start we will just use shop_id and item_id
X = train.drop(columns=['date','item_price','item_cnt_day','date_block_num']) #drop all values except shop_id and item_id
Y = train['item_cnt_day'] #what we are trying to predict
X_train, X_test, y_train, y_test = train_test_split(X, Y, shuffle=True, test_size=0.2, random_state=0) #split data into train and test sets
X_train.head()

shop_id  item_id
146198        18     5822
1576404       25    16978
2763402       21    18106
881755         7    19860
1255295       47     9574

In [4]:
# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
# Instantiate model
rf = RandomForestRegressor()
# Train the model on training data
rf.fit(X_train, y_train);

In [6]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
Y_pred=rf.predict(X_test)
print('Mean Absolute Error:', mean_absolute_error(y_test, Y_pred))
print('Mean Squared Error:', mean_squared_error(y_test, Y_pred))
print('Root Mean Squared Error:', np.sqrt(mean_squared_error(y_test, Y_pred)))
print('Coefficient of determination:', r2_score(y_test, Y_pred))


Mean Absolute Error: 0.3286098376192011
Mean Squared Error: 4.390190670924906
Root Mean Squared Error: 2.0952781846153283
Coefficient of determination: 0.29808150401995326


Our coefficient of determination is close to zero indicating that shop_id and item_id are not doing much of anything to predicted item count. While this model preformed better than our Linear Regression, these results are not very accurate. We can also see that our RMSE is 2.1 which again indicates this model is doing a poor job of predicting our item count.

#### Let's redo the model without dropping features and see which features are important
